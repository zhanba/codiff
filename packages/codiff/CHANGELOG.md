# codiff

## 0.1.1

### Patch Changes

- init publish
