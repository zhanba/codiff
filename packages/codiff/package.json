{
  "name": "codiff",
  "version": "0.1.0",
  "description": "diff algorithm of vscode",
  "keywords": [
    "diff",
    "vscode"
  ],
  "homepage": "https://github.com/zhanba/codiff#readme",
  "bugs": {
    "url": "https://github.com/zhanba/codiff/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/zhanba/codiff.git"
  },
  "license": "MIT",
  "author": "zhanba",
  "main": "dist/index.js",
  "scripts": {
    "prepublishOnly": "pnpm run build",
    "tsc": "tsc",
    "build": "rm -rf ./dist && pnpm run tsc"
  },
  "devDependencies": {
    "vitest": "^3.0.5"
  }
}