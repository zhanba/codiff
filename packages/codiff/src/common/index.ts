export * from "./arrays.js";
export * from "./arraysFind.js";
export * from "./assert.js";
export * from "./charCode.js";
export * from "./editOperation.js";
export * from "./errors.js";
export * from "./hash.js";
export * from "./lineRange.js";
export * from "./map.js";
export * from "./offsetEdit.js";
export * from "./offsetRange.js";
export * from "./position.js";
export * from "./positionToOffset.js";
export * from "./range.js";
export * from "./strings.js";
export * from "./textEdit.js";
export * from "./textLength.js";
export * from "./uint.js";
export * from "./diff/diff.js";
export * from "./diff/diffChange.js";
