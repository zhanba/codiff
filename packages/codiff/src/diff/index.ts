export * from "./linesDiffComputer.js";
export * from "./linesDiffComputers.js";
export * from "./documentDiffProvider.js";
export * from "./defaultDocumentDiffProvider.js";
