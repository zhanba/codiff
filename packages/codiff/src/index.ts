export * from "./diff/index.js";
