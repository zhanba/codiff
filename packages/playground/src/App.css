#root {
  margin: 0;
  padding: 0;
  width: 100vw;
  height: 100vh;
}

.playground {
  height: 100vh;
  display: flex;
  flex-direction: column;
}

.header {
  flex: 0 0 40px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 0 8px;
}

.toolbar {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: end;
  gap: 8px;
}

.content {
  display: flex;
  flex-direction: row;
  flex: 1 0 400px;
}

.footer {
  flex: 0 0 20px;
  font-size: 10px;
  padding: 0 12px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.links {
  display: flex;
  align-content: center;
  justify-content: space-between;
  gap: 20px;
}

.code {
  flex: 3 1 200px;
  height: calc(100vh - 60px);
  width: 100%;
  overflow: auto;
}

.result {
  flex: 1 1 200px;
  height: 100%;
  width: 100%;
  overflow: auto;
}

.cm-editor {
  height: calc(100vh - 70px);
}

input {
  width: 40px;
  margin: 0 8px;
}
