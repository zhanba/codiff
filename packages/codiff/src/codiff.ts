import { LineRange } from "./common/lineRange";
import { Range } from "./common/range";
import { splitLines } from "./common/strings";
import {
  IDocumentDiff,
  IDocumentDiffProviderOptions,
} from "./diff/documentDiffProvider";
import { linesDiffComputers } from "./diff/linesDiffComputers";
import { DetailedLineRangeMapping, RangeMapping } from "./diff/rangeMapping";

export type DiffAlgorithmName = "legacy" | "advanced";

export interface IDiffOptions extends IDocumentDiffProviderOptions {
  diffAlgorithm: DiffAlgorithmName;
}

export const defaultDiffOption: IDiffOptions = {
  ignoreTrimWhitespace: true,
  maxComputationTimeMs: 1000,
  computeMoves: false,
  extendToSubwords: false,
  diffAlgorithm: "advanced",
};

export class Codiff {
  static maxCacheSize = 100;
  private static readonly diffCache = new Map<string, IDocumentDiff>();

  protected getDiffAlgorithm(name?: DiffAlgorithmName) {
    if (name === "legacy") {
      return linesDiffComputers.getLegacy();
    }
    return linesDiffComputers.getDefault();
  }

  protected getFullRange(lines: string[]): Range {
    return new Range(
      1,
      1,
      lines.length + 1,
      lines[lines.length - 1].length + 1
    );
  }

  protected getDiffCacheKey(original: string, modified: string) {
    return `${original}-codiff-cache-key-${modified}`;
  }

  computeDiff(
    original: string,
    modified: string,
    options: IDiffOptions = defaultDiffOption
  ): IDocumentDiff {
    const originalLines = splitLines(original);
    const modifiedLines = splitLines(modified);

    // This significantly speeds up the case when the original file is empty
    if (originalLines.length === 1 && originalLines[0].length === 0) {
      if (modifiedLines.length === 1 && modifiedLines[0].length === 0) {
        return {
          changes: [],
          identical: true,
          quitEarly: false,
          moves: [],
        };
      }

      return {
        changes: [
          new DetailedLineRangeMapping(
            new LineRange(1, 2),
            new LineRange(1, modifiedLines.length + 1),
            [
              new RangeMapping(
                this.getFullRange(originalLines),
                this.getFullRange(modifiedLines)
              ),
            ]
          ),
        ],
        identical: false,
        quitEarly: false,
        moves: [],
      };
    }

    const diffAlgorithm = this.getDiffAlgorithm(options.diffAlgorithm);
    const result = diffAlgorithm.computeDiff(
      originalLines,
      modifiedLines,
      options
    );
    const diffResult: IDocumentDiff = {
      changes: result.changes,
      quitEarly: result.hitTimeout,
      identical: original === modified,
      moves: result.moves,
    };
    if (Codiff.diffCache.size > Codiff.maxCacheSize) {
      Codiff.diffCache.delete(Codiff.diffCache.keys().next().value!);
    }
    Codiff.diffCache.set(this.getDiffCacheKey(original, modified), diffResult);
    return diffResult;
  }
}
