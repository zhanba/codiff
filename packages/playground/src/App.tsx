import { useEffect, useState } from "react";
import CodeMirror from "@uiw/react-codemirror";
import "./App.css";
import { Codiff } from "codiff";
import { Change, Chunk } from "@codemirror/merge";
import { Text } from "@codemirror/state";
import { generateSimilarCode } from "./util";
import CodeMirrorMerge from "react-codemirror-merge";
import { json } from "@codemirror/lang-json";

const Original = CodeMirrorMerge.Original;
const Modified = CodeMirrorMerge.Modified;

const codiff = new Codiff();

const overrideBuild = () => {
  const buildChunks = (a: Text, b: Text): readonly Chunk[] => {
    console.log("diff compute");
    const result = codiff.computeDiff(a.toString(), b.toString());
    return result.changes.map<Chunk>((item) => {
      const getPosition = (text: Text, from: number, to: number) => {
        if (from === to) {
          // no change
          if (from === 1) {
            return { from: text.line(from).to, to: text.line(from).to };
          }
          return { from: text.line(from - 1).to, to: text.line(from - 1).to };
        }
        return { from: text.line(from).from, to: text.line(to - 1).to };
      };

      const origin = getPosition(
        a,
        item.original.startLineNumber,
        item.original.endLineNumberExclusive
      );

      const modified = getPosition(
        b,
        item.modified.startLineNumber,
        item.modified.endLineNumberExclusive
      );

      const nonNegative = (a: number) => {
        return a < 0 ? 0 : a;
      };

      const changes = (item.innerChanges ?? []).map<Change>((inner) => {
        return {
          fromA: nonNegative(
            a.line(inner.originalRange.startLineNumber).from +
              inner.originalRange.startColumn -
              1 -
              origin.from
          ),
          toA: nonNegative(
            a.line(inner.originalRange.endLineNumber).from +
              inner.originalRange.endColumn -
              1 -
              origin.from
          ),
          fromB: nonNegative(
            b.line(inner.modifiedRange.startLineNumber).from +
              inner.modifiedRange.startColumn -
              1 -
              modified.from
          ),
          toB: nonNegative(
            b.line(inner.modifiedRange.endLineNumber).from +
              inner.modifiedRange.endColumn -
              1 -
              modified.from
          ),
        };
      });

      const chunk = new Chunk(
        changes,
        origin.from,
        origin.to + 1,
        modified.from,
        modified.to + 1
      );
      return chunk;
    });
  };

  Chunk.build = buildChunks;
  Chunk.updateA = (chunks: readonly Chunk[], a: Text, b: Text) =>
    buildChunks(a, b);
  Chunk.updateB = (chunks: readonly Chunk[], a: Text, b: Text) =>
    buildChunks(a, b);
};

overrideBuild();

const doc = `one
two
three
four
five`;

const docModified = doc.replace(/t/g, "T") + "\nSix";

function App() {
  const [origin, setOrigin] = useState<string>(doc);
  const [modified, setModified] = useState<string>(docModified);
  const [diff, setDiff] = useState<string>();
  const [diffms, setDiffms] = useState<string>();

  const [maxLine, setMaxLine] = useState(20);
  const [maxDiff, setMaxDiff] = useState(10);

  const handleGenerate = () => {
    const { codeA: origin, codeB: modified } = generateSimilarCode(
      maxLine,
      maxDiff
    );
    setOrigin(origin);
    setModified(modified);
  };

  useEffect(() => {
    const start = performance.now();
    const result = codiff.computeDiff(origin, modified);
    const end = performance.now();
    setDiffms((end - start).toFixed(3));
    setDiff(JSON.stringify(result, undefined, 2));
  }, [origin, modified]);

  return (
    <div className="playground">
      <div className="header">
        <div>Codiff Playground</div>
        <div className="toolbar">
          <button onClick={() => handleGenerate()}>Generate code</button>
          <span>
            maxLine:
            <input
              value={maxLine}
              onChange={(e) => setMaxLine(parseInt(e.target.value))}
            />
          </span>
          <span>
            maxDiff:
            <input
              value={maxDiff}
              onChange={(e) => setMaxDiff(parseInt(e.target.value))}
            />
          </span>
        </div>
      </div>
      <div className="content">
        <div className="code">
          <CodeMirrorMerge theme={"dark"} destroyRerender={false}>
            <Original
              value={origin}
              onChange={(val) => {
                setOrigin(val);
              }}
              key={origin}
            />
            <Modified
              key={modified}
              value={modified}
              onChange={(val) => setModified(val)}
            />
          </CodeMirrorMerge>
        </div>
        <div className="result">
          <CodeMirror
            value={diff}
            theme={"dark"}
            className="editor"
            extensions={[json()]}
          />
        </div>
      </div>
      <div className="footer">
        <div className="links">
          <a href="https://github.com/zhanba/codiff" target="_blank">
            codiff repo
          </a>
          <a
            href="https://microsoft.github.io/monaco-editor/playground.html?source=v0.52.2#example-creating-the-diffeditor-hello-diff-world"
            target="_blank"
          >
            monaco diff editor
          </a>
          <a
            href="https://codemirror.net/try/#example=Merge%20View"
            target="_blank"
          >
            codemirror merge editor
          </a>
        </div>
        <div>diff used: {diffms} ms</div>
      </div>
    </div>
  );
}

export default App;
