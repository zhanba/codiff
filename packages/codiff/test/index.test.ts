import { describe, it, expect } from "vitest";
import { Codiff } from "../src/codiff";

describe("Codiff", () => {
  it("should create an instance of Codiff", () => {
    const codiff = new Codiff();
    expect(codiff).toBeInstanceOf(Codiff);
  });

  it("should correctly compute the diff between two strings", () => {
    const codiff = new Codiff();
    const result = codiff.computeDiff(
      `one
two
three
four
five`,
      `one
Two
Three
four
five
Six`
    );
    expect(result.changes).toEqual([
      {
        original: {
          startLineNumber: 2,
          endLineNumberExclusive: 4,
        },
        modified: {
          startLineNumber: 2,
          endLineNumberExclusive: 4,
        },
        innerChanges: [
          {
            originalRange: {
              startLineNumber: 2,
              startColumn: 1,
              endLineNumber: 2,
              endColumn: 2,
            },
            modifiedRange: {
              startLineNumber: 2,
              startColumn: 1,
              endLineNumber: 2,
              endColumn: 2,
            },
          },
          {
            originalRange: {
              startLineNumber: 3,
              startColumn: 1,
              endLineNumber: 3,
              endColumn: 2,
            },
            modifiedRange: {
              startLineNumber: 3,
              startColumn: 1,
              endLineNumber: 3,
              endColumn: 2,
            },
          },
        ],
      },
      {
        original: {
          startLineNumber: 6,
          endLineNumberExclusive: 6,
        },
        modified: {
          startLineNumber: 6,
          endLineNumberExclusive: 7,
        },
        innerChanges: [
          {
            originalRange: {
              startLineNumber: 5,
              startColumn: 5,
              endLineNumber: 5,
              endColumn: 5,
            },
            modifiedRange: {
              startLineNumber: 5,
              startColumn: 5,
              endLineNumber: 6,
              endColumn: 4,
            },
          },
        ],
      },
    ]);
  });

  it("should return an empty array when there are no differences", () => {
    const codiff = new Codiff();
    const result = codiff.computeDiff("same", "same");
    expect(result.changes).toEqual([]);
  });

  it("complex case snapshot", () => {
    const original = `for (let i722 = 0; i722 < 7; i722++) { /* loop */ }
    const fn114 = () => ['4f7omf', 'yq7ukl', 27];
    console.log('dunrt');
    const obj259 = { prop: { key: { key: 'yk4pen' } } };
    const fn719 = () => { key: 'wry0ki' };
    console.log('0t7o5');
    for (let i139 = 0; i139 < 7; i139++) { /* loop */ }
    console.log('112sqm');
    function func387(a, b) { return a + b || 0; }
    let var276 = { key: 54 };
    let var778 = '1x6xti';
    const fn765 = () => 27;
    function func544(a) { return a || 0; }
    function func170(a, b) { return a + b || 0; }
    if (var254 > 24) { /* condition */ }
    if (var252 > 47) { /* condition */ }
    if (var679 > 21) { /* condition */ }
    const obj943 = { prop: ['lharbc', 'r3iag', 90] };
    if (var818 > 6) { /* condition */ }
    const fn771 = () => { key: { key: [{ key: '0yrfhj' }, 91, '3z09h'] } };
    function func103(a) { return a || 0; }
    function func641(a) { return a || 0; }
    var71 = 16;
    function func21(a) { return a || 0; }
    if (var924 > 25) { /* condition */ }
    const obj582 = { prop: 49 };
    for (let i905 = 0; i905 < 1; i905++) { /* loop */ }
    var522 = 5;
    var349 = { key: { key: 'hnx7g' } };
    let var808 = [40, 'nrp50i', [29, 61, '2it09r']];`;

    const modified = `for (let i722 = 0; i722 < 7; i722++) { /* loop */ }
    function func374(a) { return a || 0; }
    const fn946 = () => 'o5abb2c';
    const obj256 = { prop: { key: 20 } };
    const fn114 = () => ['4f7omf', 'yq7ukl', 27];
    console.log('dunrt');
    const obj259 = { prop: { key: { key: 'yk4pen' } } };
    const obj724 = { prop: { key: 'qeabja' } };
    const fn719 = () => { key: 'wry0ki' };
    console.log('0t7o5');
    function func126(a, b) { return a + b || 0; }
    const fn158 = () => { key: 'gxyikw' };
    function func152(a, b) { return a + b || 0; }
    for(leti139=0;i139<29;i139++){/*loop*/}
    for (let i182 = 0; i182 < 2; i182++) { /* loop */ }
    functionfunc387(a,b){returna+b||49;}
    let var276 = { key: 54 };
    letvar404='1x6xti';
    const fn765 = () => 27;
    functionfunc544(a){returna||61;}
    functionfunc170(a,b){returna+b||71;}
    if (var254 > 24) { /* condition */ }
    if (var252 > 47) { /* condition */ }
    if (var499 > 23) { /* condition */ }
    if (var679 > 21) { /* condition */ }
    if (var818 > 6) { /* condition */ }
    const fn771 = () => { key: { key: [{ key: '0yrfhj' }, 91, '3z09h'] } };
    functionfunc103(a){returna||58;}
    if (var538 > 34) { /* condition */ }
    function func21(a) { return a || 0; }
    if (var924 > 25) { /* condition */ }
    for (let i905 = 0; i905 < 1; i905++) { /* loop */ }
    var522 = 5;
    var349 = { key: { key: 'hnx7g' } };
    let var808 = [40, 'nrp50i', [29, 61, '2it09r']];`;

    const codiff = new Codiff();
    const result = codiff.computeDiff(original, modified);
    expect(result.changes).toMatchSnapshot();
  });
});
