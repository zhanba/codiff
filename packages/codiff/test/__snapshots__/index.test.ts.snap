// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Codiff > complex case snapshot 1`] = `
[
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 5,
      "startLineNumber": 2,
    },
    "original": LineRange {
      "endLineNumberExclusive": 2,
      "startLineNumber": 2,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 9,
      "startLineNumber": 8,
    },
    "original": LineRange {
      "endLineNumberExclusive": 5,
      "startLineNumber": 5,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 17,
      "startLineNumber": 11,
    },
    "original": LineRange {
      "endLineNumberExclusive": 10,
      "startLineNumber": 7,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 19,
      "startLineNumber": 18,
    },
    "original": LineRange {
      "endLineNumberExclusive": 12,
      "startLineNumber": 11,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 22,
      "startLineNumber": 20,
    },
    "original": LineRange {
      "endLineNumberExclusive": 15,
      "startLineNumber": 13,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 25,
      "startLineNumber": 24,
    },
    "original": LineRange {
      "endLineNumberExclusive": 17,
      "startLineNumber": 17,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 26,
      "startLineNumber": 26,
    },
    "original": LineRange {
      "endLineNumberExclusive": 19,
      "startLineNumber": 18,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 30,
      "startLineNumber": 28,
    },
    "original": LineRange {
      "endLineNumberExclusive": 24,
      "startLineNumber": 21,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 32,
      "startLineNumber": 32,
    },
    "original": LineRange {
      "endLineNumberExclusive": 27,
      "startLineNumber": 26,
    },
  },
]
`;

exports[`Codiff > snapshot 1`] = `
[
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 5,
      "startLineNumber": 2,
    },
    "original": LineRange {
      "endLineNumberExclusive": 2,
      "startLineNumber": 2,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 9,
      "startLineNumber": 8,
    },
    "original": LineRange {
      "endLineNumberExclusive": 5,
      "startLineNumber": 5,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 17,
      "startLineNumber": 11,
    },
    "original": LineRange {
      "endLineNumberExclusive": 10,
      "startLineNumber": 7,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 19,
      "startLineNumber": 18,
    },
    "original": LineRange {
      "endLineNumberExclusive": 12,
      "startLineNumber": 11,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 22,
      "startLineNumber": 20,
    },
    "original": LineRange {
      "endLineNumberExclusive": 15,
      "startLineNumber": 13,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 25,
      "startLineNumber": 24,
    },
    "original": LineRange {
      "endLineNumberExclusive": 17,
      "startLineNumber": 17,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 26,
      "startLineNumber": 26,
    },
    "original": LineRange {
      "endLineNumberExclusive": 19,
      "startLineNumber": 18,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 30,
      "startLineNumber": 28,
    },
    "original": LineRange {
      "endLineNumberExclusive": 24,
      "startLineNumber": 21,
    },
  },
  DetailedLineRangeMapping {
    "innerChanges": [
      RangeMapping {
        "modifiedRange": [Circular],
        "originalRange": [Circular],
      },
    ],
    "modified": LineRange {
      "endLineNumberExclusive": 32,
      "startLineNumber": 32,
    },
    "original": LineRange {
      "endLineNumberExclusive": 27,
      "startLineNumber": 26,
    },
  },
]
`;
